### PR TITLE
Cldc 1671 deactivate scheme

### PR DESCRIPTION
**Pull request by natdeanlewissoftwire**
ticket: https://digital.dclg.gov.uk/jira/browse/CLDC-1671
tested locally
risk of conflict with simultaneous location deactivation work by @kosiakkatrina but we've been communicating to ameliorate this
no docs required

This is quite a large ticket, so this PR doesn't include all the functionality yet - updating logs will be done separately

deactivation:
image

confirmation:
image

success: 
image

### CODE REVIEW

File: app/views/schemes/show.html.erb [hunk 1]
```diff
@@ -15,7 +15,7 @@
 <h2 class="govuk-visually-hidden">Scheme</h2>
 
 <%= govuk_summary_list do |summary_list| %>
-  <% @scheme.display_attributes.each do |attr| %>
+  <% display_attributes(@scheme).each do |attr| %>
```
Comment (kosiakkatrina) [inline app/views/schemes/show.html.erb]: Nice!

Comment (natdeanlewissoftwire) [reply app/views/schemes/show.html.erb]: Thanks - stole this from your PR 👀

Comment (kosiakkatrina) [reply app/views/schemes/show.html.erb]: I should've updated it really 👀

File: app/controllers/schemes_controller.rb [hunk 2]
```diff
@@ -251,12 +272,55 @@ def find_resource
...
+      collection_start_date = FormHandler.instance.current_collection_start_date
+
+      if [day, month, year].any?(&:blank?)
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.not_entered"))
+      elsif !Date.valid_date?(year.to_i, month.to_i, day.to_i)
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.invalid"))
+      elsif !Date.new(year.to_i, month.to_i, day.to_i).between?(collection_start_date, Date.new(2200, 1, 1))
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.out_of_range", date: collection_start_date.to_formatted_s(:govuk_date)))
+      end
+    end
```
Comment (jamesrose) [inline app/controllers/schemes_controller.rb]: These validations are likely better suited to the model. Was there a reason for having them here?

File: app/helpers/schemes_helper.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      { name: "Status", value: scheme.status },
+    ]
+
+    if scheme.arrangement_type_same?
+      base_attributes.delete({ name: "Organisation providing support", value: scheme.managing_organisation&.name })
+    end
+    base_attributes
+  end
+
+  def availability_text(scheme)
```
Comment (jamesrose) [inline app/helpers/schemes_helper.rb]: I don't see this method used. Is it?

File: app/views/schemes/show.html.erb [hunk 4]
```diff
@@ -15,12 +15,20 @@
...
     <% end %>
   <% end %>
 <% end %>
+
+<% if FeatureToggle.scheme_toggle_enabled? %>
+  <% if @scheme.status == :active %>
+    <%= govuk_button_link_to "Deactivate this scheme", scheme_deactivate_path(@scheme), warning: true %>
+  <% else %>
+    <%= govuk_button_link_to "Reactivate this scheme", scheme_reactivate_path(@scheme) %>
+  <% end %><% end %>
```
Comment (jamesrose) [inline app/views/schemes/show.html.erb]:
```suggestion
<% if FeatureToggle.scheme_toggle_enabled? %>
  <% if @scheme.active? %>
    <%= govuk_button_link_to "Deactivate this scheme", scheme_deactivate_path(@scheme), warning: true %>
  <% else %>
    <%= govuk_button_link_to "Reactivate this scheme", scheme_reactivate_path(@scheme) %>
  <% end %>
<% end %>
```

Shall we add a helper method for this?

File: app/views/schemes/toggle_active_confirm.html.erb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<% title = "#{action.humanize} #{@scheme.service_name}" %>
+<% content_for :title, title %>
+<%= form_with model: @scheme, url: scheme_deactivate_path(@scheme), method: "patch", local: true do |f| %>
+  <% content_for :before_content do %>
+    <%= govuk_back_link(href: :back) %>
+  <% end %>
+  <h1 class="govuk-heading-l">
+    <span class="govuk-caption-l"><%= @scheme.service_name %></span>
+    <%= "This change will affect #{@scheme.lettings_logs.count} logs" %>
```
Comment (jamesrose) [inline app/views/schemes/toggle_active_confirm.html.erb]:
```suggestion
    This change will affect <%= @scheme.lettings_logs.count %> logs
```

File: app/views/schemes/toggle_active.html.erb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        <%= f.govuk_radio_button :deactivation_date_type,
+                                 "default",
+                                 label: { text: "From the start of the current collection period (#{collection_start_date.to_formatted_s(:govuk_date)})" } %>
+        <%= f.govuk_radio_button :deactivation_date_type,
+                                 "other",
+                                 label: { text: "For tenancies starting after a certain date" },
+                                 **basic_conditional_html_attributes({ "deactivation_date" => %w[other] }, "scheme") do %>
+          <%= f.govuk_date_field :deactivation_date,
+                                 legend: { text: "Date", size: "m" },
+                                 hint: { text: "For example, 27 3 2008" },
```
Comment (jamesrose) [inline app/views/schemes/toggle_active.html.erb]:
```suggestion
                                 hint: { text: "For example, 27 3 2022" },
```

This decade 😄

File: app/controllers/schemes_controller.rb (hunk 2 above)
Comment (natdeanlewissoftwire) [reply app/controllers/schemes_controller.rb]: Nope, I'll move them across 👍

File: app/helpers/schemes_helper.rb (hunk 3 above)
Comment (natdeanlewissoftwire) [reply app/helpers/schemes_helper.rb]: Ah thanks good spot, that's now redundant

File: app/models/scheme.rb [hunk 7]
```diff
@@ -230,4 +209,40 @@ def validate_confirmed
...
+
+  def available_from
+    created_at
+  end
+
+  def status
+    return :active if deactivation_date.blank?
+    return :deactivating_soon if Time.zone.now < deactivation_date
+    return :deactivated if Time.zone.now >= deactivation_date
+  end
```
Comment (ghost) [inline app/models/scheme.rb:223]:
```
  def status
    return :active if deactivation_date.blank?
    return :deactivating_soon if Time.zone.now < deactivation_date
    :deactivated
  end
```

Comment (natdeanlewissoftwire) [reply app/models/scheme.rb:223]: Cheers :)

File: app/controllers/schemes_controller.rb [hunk 8]
```diff
@@ -21,6 +21,32 @@ def show
...
+  end
+
+  def deactivate_confirm
+    render "toggle_active_confirm", locals: { action: "deactivate", deactivation_date: params[:deactivation_date], deactivation_date_type: params[:deactivation_date_type] }
+  end
+
+  def deactivate
+    prepare_for_validation
```
Comment (jamesrose) [inline app/controllers/schemes_controller.rb]: No need to extract these — it's clear if everything is in one place.

File: app/controllers/schemes_controller.rb [hunk 9]
```diff
@@ -251,12 +278,40 @@ def find_resource
...
+      flash[:notice] = "#{@scheme.service_name} has been deactivated"
+    end
+    redirect_to scheme_details_path(@scheme)
+  end
+
+  def deactivation_date
+    return if params[:scheme].blank?
+
+    collection_start_date = FormHandler.instance.current_collection_start_date
+    return collection_start_date if params[:scheme][:deactivation_date_type] == "default"
```
Comment (jamesrose) [inline app/controllers/schemes_controller.rb]:
```suggestion
    if params[:scheme][:deactivation_date_type] == "default"
      return FormHandler.instance.current_collection_start_date
    end
```

File: app/models/scheme.rb [hunk 10]
```diff
@@ -230,4 +210,36 @@ def validate_confirmed
...
+
+    :deactivated
+  end
+
+  def active?
+    status == :active
+  end
+
+  def implicit_run_deactivation_validations
+    deactivation_date.present? || @run_deactivation_validations
```
Comment (kosiakkatrina) [inline app/models/scheme.rb:234]: Would this work instead? Wouldn't need to pass run_deactivation_validations parameter if so

`deactivation_date.present? || :deactivation_date_type_changed?`

File: spec/features/schemes_spec.rb [hunk 11]
```diff
@@ -779,7 +779,7 @@
 
               it "allows to deactivate a location" do
                 click_link("Deactivate this location")
-                expect(page).to have_current_path("/schemes/#{scheme.id}/locations/#{location.id}/deactivate")
+                expect(page).to have_current_path("/schemes/#{scheme.id}/locations/#{location.id}/new-deactivation")
```
Comment (ghost) [inline spec/features/schemes_spec.rb:782]: not super important but a more RESTful endpoint would be `/schemes/#{scheme.id}/locations/#{location.id}/deactivation/new`

File: app/controllers/locations_controller.rb [hunk 12]
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
...
+      return params[:location][:deactivation_date]
+    end
 
     day = params[:location]["deactivation_date(3i)"]
     month = params[:location]["deactivation_date(2i)"]
     year = params[:location]["deactivation_date(1i)"]
+    return nil if [day, month, year].any?(&:blank?)
 
-    Date.new(year.to_i, month.to_i, day.to_i)
+    Time.utc(year.to_i, month.to_i, day.to_i) if Date.valid_date?(year.to_i, month.to_i, day.to_i)
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]:
```suggestion
    Time.local(year.to_i, month.to_i, day.to_i) if Date.valid_date?(year.to_i, month.to_i, day.to_i)
```

Always use local times. Rails does the conversion automatically.

File: app/controllers/locations_controller.rb [hunk 13]
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
...
+    elsif params[:location][:deactivation_date_type] == "default"
+      return FormHandler.instance.current_collection_start_date
+    elsif params[:location][:deactivation_date].present?
+      return params[:location][:deactivation_date]
+    end
 
     day = params[:location]["deactivation_date(3i)"]
     month = params[:location]["deactivation_date(2i)"]
     year = params[:location]["deactivation_date(1i)"]
+    return nil if [day, month, year].any?(&:blank?)
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb:204]:
```suggestion
    return if [day, month, year].any?(&:blank?)
```

File: app/controllers/locations_controller.rb [hunk 14]
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
   end
 
   def confirm_deactivation
-    if @location.update(deactivation_date: params[:location][:deactivation_date])
-      flash[:notice] = "#{@location.name || @location.postcode} has been deactivated"
+    if @location.update!(deactivation_date: @location.deactivation_date)
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]: IMO I don't see much benefit on extracting these methods.

File: spec/features/schemes_spec.rb (hunk 11 above)
Comment (natdeanlewissoftwire) [reply spec/features/schemes_spec.rb:782]: That is a bit nicer, but I think `new-deactivation,` `deactivate-confirm` and `deactivate` are ok for MVP

Comment (natdeanlewissoftwire) [reply spec/features/schemes_spec.rb:782]: There are other places we could make endpoints more RESTful too, I'll make a ticket for the future if it becomes a priority

File: app/models/scheme.rb (hunk 10 above)
Comment (natdeanlewissoftwire) [reply app/models/scheme.rb:234]: I don't think so - deactivation date type wouldn't change if a user resubmits "other" with an empty deactivation_date, but we want the validation to still trigger in that case.

Comment (natdeanlewissoftwire) [reply app/models/scheme.rb:234]: I'm not sure what that _changed? method is - do you have a link to some docs?

**Review by kosiakkatrina — APPROVED** [review]

**Pull request merged**